client/src/main/java/org/eclipse/hono/client/SendMessageSampler.java

```diff
@@ -0,0 +1,141 @@
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.client;
+
+/**
+ * An interface for sampling <em>send message</em> operations.
+ */
+public interface SendMessageSampler {
+
+    /**
+     * A factory for creating samplers.
+     */
+    interface Factory {
+
+        /**
+         * Create a new sampler.
+         *
+         * @param messageType The message type to create a sampler for.
+         * @return A new sampler.
+         */
+        SendMessageSampler create(String messageType);
+
+    }
+
+    /**
+     * Get a default, no-op implementation.
+     *
+     * @return A no-op implementation, never returns {@code null}.
+     */
+    static SendMessageSampler.Factory noop() {
+        return Noop.FACTORY;
+    }
+
+    /**
+     * Get a default, no-op implementation.
+     *
+     * @param messageType The message type to create the instance for.
+     * @return A no-op implementation, never returns {@code null}.
+     */
+    static SendMessageSampler noop(final String messageType) {
+        return noop().create(messageType);
+    }
+
+    /**
+     * A default no-op implementations.
+     */
+    class Noop {
+
+        private static final Sample SAMPLE = new Sample() {
+
+            @Override
+            public void accepted() {
+            }
+
+            @Override
+            public void rejected(final int errorCode) {
+            }
+
+            @Override
+            public void timeout() {
+            }
+
+        };
+
+        private static final SendMessageSampler SAMPLER = new SendMessageSampler() {
+
+            @Override
+            public Sample start(final String tenantId) {
+                return SAMPLE;
+            }
+
+            @Override
+            public void queueFull(final String tenantId) {
+            }
+
+        };
+
+        private static final Factory FACTORY = new Factory() {
+
+            @Override
+            public SendMessageSampler create(final String messageType) {
+                return SAMPLER;
+            }
+
+        };
+
+        private Noop() {
+        }
+
+    }
+
+    /**
+     * An active sample instance.
+     */
+    interface Sample {
+
+        /**
+         * Call when the message was accepted by the remote peer.
+         */
+        void accepted();
+
+        /**
+         * Call when the message was rejected by the remote peer.
+         *
+         * @param errorCode The rejection code. Expected to be a valid HTTP error code.
+         */
+        void rejected(int errorCode);
```

Comment:
IMHO *relreased* would also be an interesting state to record ....

Reply:
That is what the error code was intended to. `Released` would translate into `503`. The again `Modified` would translate into the same status code.

Maybe using the actual name (`Rejected`, …) is a better approach here?!

Reply:
Maybe we can combine both *accepted()* and *rejected()* into one method, e.g. *completed(String outcome)* which takes the delivery state as an argument?